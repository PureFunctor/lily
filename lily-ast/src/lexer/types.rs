//! Type definitions for the lexer.

/// The kinds of comments.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum CommentK {
    Block,
    Line,
}

/// The kinds of identifiers.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum IdentifierK {
    Ado,
    Case,
    Do,
    Else,
    If,
    In,
    Let,
    Lower,
    Of,
    Then,
    Upper,
}

/// The kinds of brackets.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum DelimiterK {
    Round,
    Square,
    Brace,
}

/// The kinds of operators.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum OperatorK {
    ArrowLeft,
    ArrowRight,
    Backslash,
    Bang,
    Colon,
    Comma,
    Equal,
    GreaterThan,
    LessThan,
    Period,
    Pipe,
    Question,
    Source,
    Underscore,
}

/// The kinds of digits.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum DigitK {
    Float,
    Int,
}

/// The kinds of invalid tokens.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum UnknownK {
    UnfinishedComment,
    UnfinishedFloat,
    UnknownToken,
}

/// The kinds of layout tokens.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum LayoutK {
    Begin,
    End,
    Separator,
}

/// The kind of a token.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum TokenK {
    CloseDelimiter(DelimiterK),
    Comment(CommentK),
    Digit(DigitK),
    Identifier(IdentifierK),
    Layout(LayoutK),
    OpenDelimiter(DelimiterK),
    Operator(OperatorK),
    Unknown(UnknownK),
    Whitespace,
}

impl TokenK {
    /// Returns `true` if the token is irrelevant for layout.
    pub fn is_annotation(&self) -> bool {
        matches!(self, TokenK::Comment(_) | TokenK::Whitespace)
    }

    /// Returns `true` if the token is relevant for layout.
    pub fn is_syntax(&self) -> bool {
        !self.is_annotation()
    }
}

/// A token produced by the lexer.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Token {
    /// The beginning byte offset.
    pub begin: usize,
    /// The ending byte offset.
    pub end: usize,
    /// The kind of the token.
    pub kind: TokenK,
    /// The "layout depth" of the token.
    pub depth: usize,
}

impl Token {
    /// Returns `true` if the token is irrelevant for layout.
    pub fn is_annotation(&self) -> bool {
        self.kind.is_annotation()
    }

    /// Returns `true` if the token is relevant for layout.
    pub fn is_syntax(&self) -> bool {
        self.kind.is_syntax()
    }

    /// Sets a new value for a [`Token`]'s `depth` field.
    pub fn with_depth(mut self, depth: usize) -> Self {
        self.depth = depth;
        self
    }
}
