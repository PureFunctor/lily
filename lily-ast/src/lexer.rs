//! Top-level API for the lexer.
//!
//! # Basic Usage
//!
//! ```rust
//! use lily_ast::lexer::lex;
//!
//! let tokens = lex("a = 0\nb = 0");
//!
//! assert_eq!(tokens.len(), 13);
//! ```
pub use self::{cursor::tokenize, layout::with_layout, types::Token};

pub mod cursor;
pub mod layout;
pub mod types;

/// Creates a stream of tokens from a source file.
pub fn lex(source: &str) -> Vec<Token> {
    with_layout(source, tokenize(source))
}

#[cfg(test)]
mod tests {
    use pretty_assertions::assert_eq;

    use super::{
        lex,
        types::{LayoutK, TokenK},
    };

    const SOURCE: &str = r"Identity : Type -> Type
Identity a ?
  _ : a -> Identity a

Equal : Type -> Type -> Boolean
Equal a b !
  _ : a -> a -> True
  _ : a -> b -> False

Eq : Type -> Constraint
Eq a |
  eq : a -> a -> Boolean

head : List a -> Maybe a
head xs = case xs of
  Cons x _ -> Just x
  Nil      -> Nothing

main : Effect Unit
main = do
  log message
  log message
  attempt do
    log message
    log message

ofCollapse : Int
ofCollapse =
  case
    do _ <- pure 0
       pure 1
  of
    Just x -> x
    Nothing -> 0

lambdaMask : List a -> Maybe a
lambdaMask xs = case xs of
  Cons x _ if (\_ -> true) x ->
    Just x
  _ ->
    Nothing

arrowFinishDo : List a -> Maybe a
arrowFinishDo xs = case xs of
  Cons x _ if do true ->
    Just x
  _ ->
    Nothing

conditionalDo : Effect Unit
conditionalDo = do
  log something
  if do true then do
    log something
  else do
    log something

letIn : Int
letIn =
  let
    x : Int
    x = 1

    y : Int
    y = 1
  in
    x + y

adoIn : Int
adoIn = ado
  x <- pure 1
  y <- pure 1
  let
    a : Int
    a = let b = c in d

    e : Int
    e = let f = g in h
  in x + y

adoLet : Effect Unit
adoLet = do
  logShow $ x + y
  let
    x : Int
    x = 1

    y : Int
    y = 1
  logShow $ x + y
";

    #[test]
    fn ascending_position() {
        let tokens = lex(SOURCE);
        for window in tokens.windows(2) {
            assert!(window[0].begin <= window[1].begin);
            assert!(window[0].end <= window[1].end);
        }
    }

    #[test]
    fn basic_layout_test() {
        let mut actual = String::new();
        let expected = r"Identity : Type -> Type;0
Identity a ?{1
  _ : a -> Identity a;1}1;0

Equal : Type -> Type -> Boolean;0
Equal a b !{1
  _ : a -> a -> True;1
  _ : a -> b -> False;1}1;0

Eq : Type -> Constraint;0
Eq a |{1
  eq : a -> a -> Boolean;1}1;0

head : List a -> Maybe a;0
head xs = case xs of{1
  Cons x _ -> Just x;1
  Nil      -> Nothing;1}1;0

main : Effect Unit;0
main = do{1
  log message;1
  log message;1
  attempt do{2
    log message;2
    log message;2}2;1}1;0

ofCollapse : Int;0
ofCollapse =
  case
    do{1 _ <- pure 0;1
       pure 1;1}1
  of{1
    Just x -> x;1
    Nothing -> 0;1}1;0

lambdaMask : List a -> Maybe a;0
lambdaMask xs = case xs of{1
  Cons x _ if (\_ -> true) x ->
    Just x;1
  _ ->
    Nothing;1}1;0

arrowFinishDo : List a -> Maybe a;0
arrowFinishDo xs = case xs of{1
  Cons x _ if do{2 true;2}2 ->
    Just x;1
  _ ->
    Nothing;1}1;0

conditionalDo : Effect Unit;0
conditionalDo = do{1
  log something;1
  if do{2 true;2}2 then do{2
    log something;2}2
  else do{2
    log something;2}2;1}1;0

letIn : Int;0
letIn =
  let{1
    x : Int;1
    x = 1;1

    y : Int;1
    y = 1;1}1
  in
    x + y;0

adoIn : Int;0
adoIn = ado{1
  x <- pure 1;1
  y <- pure 1;1
  let{2
    a : Int;2
    a = let{3 b = c;3}3 in d;2

    e : Int;2
    e = let{3 f = g;3}3 in h;2}2;1}1
  in x + y;0

adoLet : Effect Unit;0
adoLet = do{1
  logShow $ x + y;1
  let{2
    x : Int;2
    x = 1;2

    y : Int;2
    y = 1;2}2;1
  logShow $ x + y;1}1;0
";

        let mut collected_whitespace = vec![];

        for token in lex(SOURCE) {
            if let TokenK::Whitespace = token.kind {
                collected_whitespace.push(token);
            } else if let TokenK::Layout(layout) = token.kind {
                match layout {
                    LayoutK::Begin => actual.push_str(format!("{{{}", token.depth).as_str()),
                    LayoutK::End => actual.push_str(format!("}}{}", token.depth).as_str()),
                    LayoutK::Separator => actual.push_str(format!(";{}", token.depth).as_str()),
                }
            } else {
                while let Some(whitespace) = collected_whitespace.pop() {
                    actual.push_str(&SOURCE[whitespace.begin..whitespace.end]);
                }
                actual.push_str(&SOURCE[token.begin..token.end]);
            }
        }

        actual.push('\n');
        print!("{}", actual);
        assert_eq!(actual, expected);
    }
}
