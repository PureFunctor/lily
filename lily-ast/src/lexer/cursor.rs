use std::str::Chars;

use unicode_categories::UnicodeCategories;

use super::types::{DelimiterK, DigitK, IdentifierK, OperatorK, Token, TokenK, UnknownK};

#[derive(Debug, Clone)]
pub(crate) struct Cursor<'a> {
    length: usize,
    source: &'a str,
    chars: Chars<'a>,
}

const EOF_CHAR: char = '\0';

impl<'a> Cursor<'a> {
    pub(crate) fn new(source: &'a str) -> Self {
        Self {
            length: source.len(),
            source,
            chars: source.chars(),
        }
    }

    fn is_eof(&self) -> bool {
        self.chars.as_str().is_empty()
    }

    fn consumed(&self) -> usize {
        self.length - self.chars.as_str().len()
    }

    fn peek_1(&mut self) -> char {
        self.chars.clone().next().unwrap_or(EOF_CHAR)
    }

    fn peek_2(&mut self) -> char {
        let mut chars = self.chars.clone();
        chars.next();
        chars.next().unwrap_or(EOF_CHAR)
    }

    fn take(&mut self) -> char {
        self.chars.next().unwrap_or(EOF_CHAR)
    }

    fn take_while(&mut self, predicate: impl Fn(char) -> bool) {
        while predicate(self.peek_1()) && !self.is_eof() {
            self.take();
        }
    }
}

impl<'a> Cursor<'a> {
    pub(crate) fn take_token(&mut self) -> Token {
        let comment_begin = self.consumed();
        loop {
            match (self.peek_1(), self.peek_2()) {
                ('-', '-') => {
                    self.take_while(|c| c != '\n');
                }
                ('{', '-') => loop {
                    if self.is_eof() {
                        break;
                    } else if self.peek_1() == '-' && self.peek_2() == '}' {
                        self.take();
                        self.take();
                        break;
                    } else {
                        self.take();
                    }
                },
                (i, _) if i.is_whitespace() => {
                    self.take_while(|c| c.is_whitespace());
                }
                _ => break,
            }
        }
        let comment_end = self.consumed();
        let begin = self.consumed();
        let kind = match self.take() {
            // Open Parentheses
            '(' => TokenK::OpenDelimiter(DelimiterK::Round),
            '[' => TokenK::OpenDelimiter(DelimiterK::Square),
            '{' => TokenK::OpenDelimiter(DelimiterK::Brace),
            // Close Parentheses
            ')' => TokenK::CloseDelimiter(DelimiterK::Round),
            ']' => TokenK::CloseDelimiter(DelimiterK::Square),
            '}' => TokenK::CloseDelimiter(DelimiterK::Brace),
            // Built-in Symbols
            ',' => TokenK::Operator(OperatorK::Comma),
            '\\' => TokenK::Operator(OperatorK::Backslash),
            // Identifiers
            initial if initial.is_letter_lowercase() || initial == '_' && self.peek_1() == '_' => {
                self.take_while(|c| c.is_letter() || c.is_number() || "'_".contains(c));
                let end = self.consumed();
                TokenK::Identifier(match &self.source[begin..end] {
                    "ado" => IdentifierK::Ado,
                    "case" => IdentifierK::Case,
                    "do" => IdentifierK::Do,
                    "else" => IdentifierK::Else,
                    "if" => IdentifierK::If,
                    "in" => IdentifierK::In,
                    "let" => IdentifierK::Let,
                    "of" => IdentifierK::Of,
                    "then" => IdentifierK::Then,
                    _ => IdentifierK::Lower,
                })
            }
            initial if initial.is_letter_uppercase() => {
                self.take_while(|c| c.is_letter() || c.is_number() || "'_".contains(c));
                TokenK::Identifier(IdentifierK::Upper)
            }
            // Compound Symbols
            initial if initial.is_symbol() || initial.is_punctuation() => {
                self.take_while(|c| !"(){}[]".contains(c) && (c.is_symbol() || c.is_punctuation()));
                let end = self.consumed();
                TokenK::Operator(match &self.source[begin..end] {
                    "->" => OperatorK::ArrowRight,
                    "<-" => OperatorK::ArrowLeft,
                    "=" => OperatorK::Equal,
                    ":" => OperatorK::Colon,
                    "." => OperatorK::Period,
                    "|" => OperatorK::Pipe,
                    "?" => OperatorK::Question,
                    "!" => OperatorK::Bang,
                    "_" => OperatorK::Underscore,
                    "<" => OperatorK::LessThan,
                    ">" => OperatorK::GreaterThan,
                    _ => OperatorK::Source,
                })
            }
            // Digits
            initial if initial.is_ascii_digit() => {
                self.take_while(|c| c.is_ascii_digit());
                if self.peek_1() == '.' {
                    // 1..
                    if self.peek_2() == '.' {
                        TokenK::Digit(DigitK::Int)
                    // 1.2
                    } else if self.peek_2().is_ascii_digit() {
                        self.take();
                        self.take_while(|c| c.is_ascii_digit());
                        TokenK::Digit(DigitK::Float)
                    // 1.
                    } else {
                        self.take();
                        TokenK::Unknown(UnknownK::UnfinishedFloat)
                    }
                } else {
                    TokenK::Digit(DigitK::Int)
                }
            }
            // End of file
            '\0' => TokenK::Unknown(UnknownK::EndOfFile),
            // Unknown Token
            _ => TokenK::Unknown(UnknownK::UnknownToken),
        };
        let end = self.consumed();
        let depth = 0;
        Token {
            comment_begin,
            comment_end,
            begin,
            end,
            kind,
            depth,
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::lexer::types::UnknownK;

    use super::{Cursor, DigitK, IdentifierK, OperatorK, Token, TokenK};
    use pretty_assertions::assert_eq;

    #[test]
    fn double_period_after_int() {
        let source = "1..2";
        let mut cursor = Cursor::new(source);
        assert_eq!(
            cursor.take_token(),
            Token {
                comment_begin: 0,
                comment_end: 0,
                begin: 0,
                end: 1,
                kind: TokenK::Digit(DigitK::Int),
                depth: 0,
            }
        );
        assert_eq!(
            cursor.take_token(),
            Token {
                comment_begin: 1,
                comment_end: 1,
                begin: 1,
                end: 3,
                kind: TokenK::Operator(OperatorK::Source),
                depth: 0,
            }
        );
        assert_eq!(
            cursor.take_token(),
            Token {
                comment_begin: 3,
                comment_end: 3,
                begin: 3,
                end: 4,
                kind: TokenK::Digit(DigitK::Int),
                depth: 0,
            }
        );
        assert_eq!(
            cursor.take_token(),
            Token {
                comment_begin: 4,
                comment_end: 4,
                begin: 4,
                end: 4,
                kind: TokenK::Unknown(UnknownK::EndOfFile),
                depth: 0,
            }
        );
    }

    #[test]
    fn block_comment_in_between() {
        let source = "1 {-hello-} 2";
        let mut cursor = Cursor::new(source);
        assert_eq!(
            cursor.take_token(),
            Token {
                comment_begin: 0,
                comment_end: 0,
                begin: 0,
                end: 1,
                kind: TokenK::Digit(DigitK::Int),
                depth: 0,
            }
        );
        assert_eq!(
            cursor.take_token(),
            Token {
                comment_begin: 1,
                comment_end: 12,
                begin: 12,
                end: 13,
                kind: TokenK::Digit(DigitK::Int),
                depth: 0,
            }
        );
        assert_eq!(
            cursor.take_token(),
            Token {
                comment_begin: 13,
                comment_end: 13,
                begin: 13,
                end: 13,
                kind: TokenK::Unknown(UnknownK::EndOfFile),
                depth: 0,
            }
        );
    }

    #[test]
    fn underscore_disambiguation() {
        let source = "__";
        let mut cursor = Cursor::new(source);
        assert_eq!(
            cursor.take_token(),
            Token {
                comment_begin: 0,
                comment_end: 0,
                begin: 0,
                end: 2,
                kind: TokenK::Identifier(IdentifierK::Lower),
                depth: 0,
            }
        );
        assert_eq!(
            cursor.take_token(),
            Token {
                comment_begin: 2,
                comment_end: 2,
                begin: 2,
                end: 2,
                kind: TokenK::Unknown(UnknownK::EndOfFile),
                depth: 0,
            }
        )
    }
}
