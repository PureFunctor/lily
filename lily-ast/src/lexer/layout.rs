use super::types::{IdentifierK, LayoutK, OperatorK, Position, Token, TokenK};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum DelimiterK {
    KwAdo,
    KwCase,
    KwDo,
    KwIf,
    KwLetExpr,
    KwLetStmt,
    KwOf,
    KwThen,
    MaskLam,
    MaskPat,
    MaskRoot,
    MaskTop,
}

impl DelimiterK {
    fn is_indented(&self) -> bool {
        use DelimiterK::*;
        matches!(
            &self,
            KwAdo | KwDo | KwLetExpr | KwLetStmt | KwOf | MaskRoot | MaskTop
        )
    }
}

pub(crate) struct LayoutEngine {
    delimiters: Vec<(Position, DelimiterK)>,
    pub depth: usize,
}

impl LayoutEngine {
    pub(crate) fn new(initial_position: Position) -> Self {
        let delimiters = vec![(initial_position, DelimiterK::MaskRoot)];
        let depth = 0;
        Self { delimiters, depth }
    }

    #[inline]
    fn push_end(&mut self, tokens: &mut Vec<Token>, current_token: Token) {
        let begin = current_token.begin;
        let end = current_token.begin;
        let depth = self.depth;
        tokens.push(Token {
            comment_begin: begin,
            comment_end: begin,
            begin,
            end,
            kind: TokenK::Layout(LayoutK::Separator),
            depth,
        });
        tokens.push(Token {
            comment_begin: begin,
            comment_end: begin,
            begin,
            end,
            kind: TokenK::Layout(LayoutK::End),
            depth,
        });
        self.depth = self.depth.saturating_sub(1);
    }

    #[inline]
    fn determine_end(&self, predicate: impl Fn(&Position, &DelimiterK) -> bool) -> (usize, usize) {
        let mut take_n = self.delimiters.len();
        let mut make_n = 0;

        for (position, delimiter) in self.delimiters.iter().rev() {
            if predicate(position, delimiter) {
                take_n = take_n.saturating_sub(1);
                if delimiter.is_indented() {
                    make_n += 1;
                }
            } else {
                break;
            }
        }

        (take_n, make_n)
    }

    #[inline]
    fn add_begin(
        &mut self,
        tokens: &mut Vec<Token>,
        current_token: Token,
        next_position: Position,
        delimiter: DelimiterK,
    ) {
        let recent_indented = self
            .delimiters
            .iter()
            .rfind(|(_, delimiter)| delimiter.is_indented());

        if let Some((past, _)) = recent_indented {
            if next_position.column <= past.column {
                return;
            }
        }

        self.delimiters.push((next_position, delimiter));
        self.depth += 1;
        tokens.push(Token {
            comment_begin: current_token.end,
            comment_end: current_token.end,
            begin: current_token.end,
            end: current_token.end,
            kind: TokenK::Layout(LayoutK::Begin),
            depth: self.depth,
        });
    }

    #[inline]
    fn add_separator(
        &mut self,
        tokens: &mut Vec<Token>,
        current_token: Token,
        now_position: Position,
    ) {
        if let Some((position, delimiter)) = self.delimiters.last() {
            if delimiter.is_indented()
                && now_position.column == position.column
                && now_position.line > position.line
            {
                tokens.push(Token {
                    comment_begin: current_token.begin,
                    comment_end: current_token.begin,
                    begin: current_token.begin,
                    end: current_token.begin,
                    kind: TokenK::Layout(LayoutK::Separator),
                    depth: self.depth,
                });
                if let DelimiterK::KwOf = delimiter {
                    self.delimiters.push((now_position, DelimiterK::MaskPat));
                }
            }
        }
    }

    #[inline]
    fn add_end(&mut self, tokens: &mut Vec<Token>, current_token: Token, now_position: Position) {
        let (take_n, make_n) = self.determine_end(|position, delimiter| {
            delimiter.is_indented() && now_position.column < position.column
        });
        self.delimiters.truncate(take_n);
        for _ in 0..make_n {
            self.push_end(tokens, current_token);
        }
    }

    #[inline]
    pub(crate) fn add_layout(
        &mut self,
        tokens: &mut Vec<Token>,
        current_token: Token,
        now_position: Position,
        next_position: Position,
    ) {
        use DelimiterK::*;
        use IdentifierK::*;
        use OperatorK::*;
        use TokenK::*;

        macro_rules! with_end {
            ($predicate:expr, $($commit:literal ~ $pattern:pat $(if $guard:expr)? => $expression:expr,)+) => {
                {
                    let (take_n, make_n) = self.determine_end($predicate);
                    match &self.delimiters[..take_n] {
                        $($pattern $(if $guard)? => {
                            if $commit {
                                self.delimiters.truncate(take_n);
                                for _ in 0..make_n {
                                    self.push_end(tokens, current_token);
                                }
                            };
                            $expression
                        }),+
                    }
                }
            };
        }

        match current_token.kind {
            Operator(Bang | Pipe | Question) => {
                tokens.push(current_token.with_depth(self.depth));
                self.add_begin(tokens, current_token, next_position, MaskTop);
            }
            Identifier(Case) => {
                self.add_end(tokens, current_token, now_position);
                self.add_separator(tokens, current_token, now_position);
                self.delimiters.push((now_position, KwCase));
                tokens.push(current_token.with_depth(self.depth));
            }
            Identifier(Of) => with_end!(
                |_, delimiter| delimiter.is_indented(),
                true ~ [.., (_, KwCase)] => {
                    self.delimiters.pop();
                    tokens.push(current_token.with_depth(self.depth));
                    self.add_begin(tokens, current_token, next_position, KwOf);
                    self.delimiters.push((next_position, MaskPat));
                },
                true ~ _ => {
                    self.add_end(tokens, current_token, now_position);
                    self.add_separator(tokens, current_token, now_position);
                    tokens.push(current_token.with_depth(self.depth));
                },
            ),
            Operator(Backslash) => {
                self.add_end(tokens, current_token, now_position);
                self.add_separator(tokens, current_token, now_position);
                self.delimiters.push((now_position, MaskLam));
                tokens.push(current_token.with_depth(self.depth));
            }
            Operator(ArrowRight) => with_end!(
                |position, delimiter| {
                    match delimiter {
                        KwDo => true,
                        KwOf => false,
                        _ => {
                            delimiter.is_indented() && now_position.column <= position.column
                        },
                    }
                },
                true ~ _ => {
                    if let Some((_, KwIf)) = self.delimiters.last() {
                        self.delimiters.pop();
                    }
                    if let Some((_, MaskLam | MaskPat)) = self.delimiters.last() {
                        self.delimiters.pop();
                    }
                    tokens.push(current_token.with_depth(self.depth));
                },
            ),
            Identifier(Ado) => {
                self.add_end(tokens, current_token, now_position);
                self.add_separator(tokens, current_token, now_position);
                tokens.push(current_token.with_depth(self.depth));
                self.add_begin(tokens, current_token, next_position, KwAdo);
            }
            Identifier(Do) => {
                self.add_end(tokens, current_token, now_position);
                self.add_separator(tokens, current_token, now_position);
                tokens.push(current_token.with_depth(self.depth));
                self.add_begin(tokens, current_token, next_position, KwDo);
            }
            Identifier(If) => {
                self.add_end(tokens, current_token, now_position);
                self.add_separator(tokens, current_token, now_position);
                self.delimiters.push((now_position, KwIf));
                tokens.push(current_token.with_depth(self.depth));
            }
            Identifier(Then) => with_end!(
                |_, delimiter| delimiter.is_indented(),
                true ~ [.., (_, KwIf)] => {
                    self.delimiters.pop();
                    self.delimiters.push((now_position, KwThen));
                    tokens.push(current_token.with_depth(self.depth));
                },
                false ~ _ => {
                    self.add_end(tokens, current_token, now_position);
                    self.add_separator(tokens, current_token, now_position);
                    tokens.push(current_token.with_depth(self.depth));
                },
            ),
            Identifier(Else) => with_end!(
                |_, delimiter| delimiter.is_indented(),
                true ~ [.., (_, KwThen)] => {
                    self.delimiters.pop();
                    tokens.push(current_token.with_depth(self.depth));
                },
                false ~ _ => {
                    self.add_end(tokens, current_token, now_position);
                    self.add_separator(tokens, current_token, now_position);
                    tokens.push(current_token.with_depth(self.depth));
                },
            ),
            Identifier(Let) => {
                self.add_end(tokens, current_token, now_position);
                self.add_separator(tokens, current_token, now_position);
                tokens.push(current_token.with_depth(self.depth));
                self.add_begin(
                    tokens,
                    current_token,
                    next_position,
                    match self.delimiters.last() {
                        Some((_, KwAdo | KwDo)) => KwLetStmt,
                        _ => KwLetExpr,
                    },
                );
            }
            Identifier(In) => with_end!(
                |_, delimiter| {
                    match delimiter {
                        KwAdo | KwLetExpr => false,
                        _ => delimiter.is_indented(),
                    }
                },
                true ~ [.., (_, KwAdo | KwLetExpr)] => {
                    self.delimiters.pop();
                    self.push_end(tokens, current_token);
                    tokens.push(current_token.with_depth(self.depth));
                },
                false ~ _ => {
                    self.add_end(tokens, current_token, now_position);
                    self.add_separator(tokens, current_token, now_position);
                    tokens.push(current_token.with_depth(self.depth));
                },
            ),
            _ => {
                self.add_end(tokens, current_token, now_position);
                self.add_separator(tokens, current_token, now_position);
                tokens.push(current_token.with_depth(self.depth));
            }
        }
    }

    pub(crate) fn finalize_layout(&mut self, tokens: &mut Vec<Token>, eof_offset: usize) {
        while let Some((_, delimiter)) = self.delimiters.pop() {
            if let DelimiterK::MaskRoot = delimiter {
                tokens.push(Token {
                    comment_begin: eof_offset,
                    comment_end: eof_offset,
                    begin: eof_offset,
                    end: eof_offset,
                    kind: TokenK::Layout(LayoutK::Separator),
                    depth: self.depth,
                });
            } else if delimiter.is_indented() {
                tokens.push(Token {
                    comment_begin: eof_offset,
                    comment_end: eof_offset,
                    begin: eof_offset,
                    end: eof_offset,
                    kind: TokenK::Layout(LayoutK::Separator),
                    depth: self.depth,
                });
                tokens.push(Token {
                    comment_begin: eof_offset,
                    comment_end: eof_offset,
                    begin: eof_offset,
                    end: eof_offset,
                    kind: TokenK::Layout(LayoutK::End),
                    depth: self.depth,
                });
                self.depth = self.depth.saturating_sub(1);
            }
        }
    }
}

// #[cfg(test)]
// mod tests {
//     use pretty_assertions::assert_eq;

//     use crate::lexer::{
//         tokenize,
//         types::{LayoutK, TokenK},
//     };

//     use super::with_layout;

//     const SOURCE: &str = r"Identity : Type -> Type
// Identity a ?
//   _ : a -> Identity a

// Equal : Type -> Type -> Boolean
// Equal a b !
//   _ : a -> a -> True
//   _ : a -> b -> False

// Eq : Type -> Constraint
// Eq a |
//   eq : a -> a -> Boolean

// head : List a -> Maybe a
// head xs = case xs of
//   Cons x _ -> Just x
//   Nil      -> Nothing

// main : Effect Unit
// main = do
//   log message
//   log message
//   attempt do
//     log message
//     log message

// ofCollapse : Int
// ofCollapse =
//   case
//     do _ <- pure 0
//        pure 1
//   of
//     Just x -> x
//     Nothing -> 0

// lambdaMask : List a -> Maybe a
// lambdaMask xs = case xs of
//   Cons x _ if (\_ -> true) x ->
//     Just x
//   _ ->
//     Nothing

// arrowFinishDo : List a -> Maybe a
// arrowFinishDo xs = case xs of
//   Cons x _ if do true ->
//     Just x
//   _ ->
//     Nothing

// conditionalDo : Effect Unit
// conditionalDo = do
//   log something
//   if do true then do
//     log something
//   else do
//     log something

// letIn : Int
// letIn =
//   let
//     x : Int
//     x = 1

//     y : Int
//     y = 1
//   in
//     x + y

// adoIn : Int
// adoIn = ado
//   x <- pure 1
//   y <- pure 1
//   let
//     a : Int
//     a = let b = c in d

//     e : Int
//     e = let f = g in h
//   in x + y

// adoLet : Effect Unit
// adoLet = do
//   logShow $ x + y
//   let
//     x : Int
//     x = 1

//     y : Int
//     y = 1
//   logShow $ x + y
// ";

//     #[test]
//     fn ascending_position() {
//         let tokens = with_layout(SOURCE, tokenize(SOURCE));
//         for window in tokens.windows(2) {
//             assert!(window[0].begin <= window[1].begin);
//             assert!(window[0].end <= window[1].end);
//         }
//     }

//     #[test]
//     fn basic_layout_test() {
//         let mut actual = String::new();
//         let expected = r"Identity : Type -> Type;0
// Identity a ?{1
//   _ : a -> Identity a;1}1;0

// Equal : Type -> Type -> Boolean;0
// Equal a b !{1
//   _ : a -> a -> True;1
//   _ : a -> b -> False;1}1;0

// Eq : Type -> Constraint;0
// Eq a |{1
//   eq : a -> a -> Boolean;1}1;0

// head : List a -> Maybe a;0
// head xs = case xs of{1
//   Cons x _ -> Just x;1
//   Nil      -> Nothing;1}1;0

// main : Effect Unit;0
// main = do{1
//   log message;1
//   log message;1
//   attempt do{2
//     log message;2
//     log message;2}2;1}1;0

// ofCollapse : Int;0
// ofCollapse =
//   case
//     do{1 _ <- pure 0;1
//        pure 1;1}1
//   of{1
//     Just x -> x;1
//     Nothing -> 0;1}1;0

// lambdaMask : List a -> Maybe a;0
// lambdaMask xs = case xs of{1
//   Cons x _ if (\_ -> true) x ->
//     Just x;1
//   _ ->
//     Nothing;1}1;0

// arrowFinishDo : List a -> Maybe a;0
// arrowFinishDo xs = case xs of{1
//   Cons x _ if do{2 true;2}2 ->
//     Just x;1
//   _ ->
//     Nothing;1}1;0

// conditionalDo : Effect Unit;0
// conditionalDo = do{1
//   log something;1
//   if do{2 true;2}2 then do{2
//     log something;2}2
//   else do{2
//     log something;2}2;1}1;0

// letIn : Int;0
// letIn =
//   let{1
//     x : Int;1
//     x = 1;1

//     y : Int;1
//     y = 1;1}1
//   in
//     x + y;0

// adoIn : Int;0
// adoIn = ado{1
//   x <- pure 1;1
//   y <- pure 1;1
//   let{2
//     a : Int;2
//     a = let{3 b = c;3}3 in d;2

//     e : Int;2
//     e = let{3 f = g;3}3 in h;2}2;1}1
//   in x + y;0

// adoLet : Effect Unit;0
// adoLet = do{1
//   logShow $ x + y;1
//   let{2
//     x : Int;2
//     x = 1;2

//     y : Int;2
//     y = 1;2}2;1
//   logShow $ x + y;1}1;0
// ";

//         for token in with_layout(SOURCE, tokenize(SOURCE)) {
//             if let TokenK::Layout(layout) = token.kind {
//                 match layout {
//                     LayoutK::Begin => actual.push_str(format!("{{{}", token.depth).as_str()),
//                     LayoutK::End => actual.push_str(format!("}}{}", token.depth).as_str()),
//                     LayoutK::Separator => actual.push_str(format!(";{}", token.depth).as_str()),
//                 }
//             } else {
//                 actual.push_str(&SOURCE[token.comment_begin..token.comment_end]);
//                 actual.push_str(&SOURCE[token.begin..token.end]);
//             }
//         }

//         assert_eq!(actual, expected);
//     }
// }
