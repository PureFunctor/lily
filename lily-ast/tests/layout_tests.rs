use std::{env, fs, path::Path};

use lily_ast::lexer::{
    lex,
    types::{LayoutK, TokenK},
};

fn lex_print(source: &str) -> String {
    let tokens = lex(source);
    let mut buffer = String::new();
    for token in tokens {
        if let TokenK::Layout(layout) = token.kind {
            match layout {
                LayoutK::Begin => buffer.push_str(format!("{{{}", token.depth).as_str()),
                LayoutK::End => buffer.push_str(format!("}}{}", token.depth).as_str()),
                LayoutK::Separator => buffer.push_str(format!(";{}", token.depth).as_str()),
            }
        } else {
            buffer.push_str(&source[token.comment_begin..token.comment_end]);
            buffer.push_str(&source[token.begin..token.end]);
        }
    }
    buffer
}

#[test]
fn golden_layout_tests() -> anyhow::Result<()> {
    let layout_files = Path::new(&env::var("CARGO_MANIFEST_DIR")?)
        .join("tests")
        .join("layout_tests");

    for file_entry in fs::read_dir(layout_files.as_path())? {
        let file_entry = file_entry?;
        let file_contents = fs::read_to_string(file_entry.path())?;
        insta::assert_snapshot!(lex_print(&file_contents));
    }

    Ok(())
}
