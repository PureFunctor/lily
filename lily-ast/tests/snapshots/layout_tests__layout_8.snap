---
source: lily-ast/tests/layout_tests.rs
expression: lex_print(&source)
---
conditionalDo : Effect Unit;0
conditionalDo = do{1
  log something;1
  if do{2 true;2}2 then do{2
    log something;2}2
  else do{2
    log something;2}2;1}1;0
